File: src/10_Bit_Manipulation/dp_optimal_selection.cpp
#include <bits/stdc++.h>
using namespace std;


// products
const int k = 3;
// days
const int n = 8;
int price[k][n] = {
  {6, 9, 5, 2, 8, 9, 1, 6},
  {8, 2, 6, 2, 7, 5, 7, 2},
  {5, 3, 9, 7, 3, 5, 1, 4}
};


// O(k^n)
int solve_rec(int s, int d)
{
  if (s == 0) return 0;
  if (d == 0) {
    // check if s is a power of 2 (namely, a single-item set)
    if ((s & (s - 1)) == 0) return price[(int)log2(s)][0];
    // s is a multy-item set, buy the cheapest product at day 0
    int best = INT_MAX;
    for (int x = 0; x < k; ++x) {
      // if x in s
      if (s & (1 << x)) best = min(best, price[x][0]);
    }
    return best;
  }

  // either we do not buy a product at day d
  int best = solve_rec(s, d - 1);
  // or buy a product x that belongs to s
  for (int x = 0; x < k; ++x) {
    if (s & (1 << x)) {
      // s ^ (1 << x) : remove x from s
      best = min(best, solve_rec(s ^ (1 << x), d - 1) + price[x][d]);
    }
  }
  return best;
}


// total[s][d]: min price of buying subset s at day d
int total[1 << k][n];


// O(nk2^k)
int solve_dp()
{
  // dp initialization
  // 0 total cost for empty subset at any day
  fill(total[0], total[0] + n, 0);
  // we can only pick one item at the 0th day
  for (int s = 1; s < (1 << k); ++s) {
    // if s is a single-item set
    // if (__builtin_popcount(s) == 1) {
    if ((s & (s - 1)) == 0) {
      total[s][0] = price[(int)log2(s)][0];
    } else {
      // if s is a multi-item set, buy the cheapest product at day 0
      int best = INT_MAX;
      for (int x = 0; x < k; ++x) {
        if (s & (1 << x)) best = min(best, price[x][0]);
      }
      total[s][0] = best;
    }
  }

  // rest of the days
  for (int d = 1; d < n; ++d) {
    for (int s = 1; s < (1 << k); ++s) {
      total[s][d] = total[s][d - 1];
      for (int x = 0; x < k; ++x) {
        // if x in s
        if (s & (1 << x)) {
          // s ^ (1 << x) : remove x from s
          total[s][d] = min(total[s][d],
                            total[s ^ (1 << x)][d - 1] + price[x][d]);
        }
      }
    }
  }
  return total[(1 << k) - 1][n - 1];
}


int main()
{
  int cost_rec = solve_rec((1 << k) - 1, n - 1);
  int cost_dp = solve_dp();
  cout << "total cost: " << cost_dp << '\n';
  assert(cost_rec == cost_dp);
}

File: src/10_Bit_Manipulation/dp_optimal_selection.py
from math import log2


price = [
  [6, 9, 5, 2, 8, 9, 1, 6],
  [8, 2, 6, 2, 7, 5, 7, 2],
  [5, 3, 9, 7, 3, 5, 1, 4]
]

# num products
k = len(price)
# num days
n = len(price[0])


# O(2^k * n * k)
# The function operates on a (s, d) state. s takes 1 << k = 2^k values and d
# takes n values. Each (s, d) state has at most k recursive calls.
def solve_rec(s, d):
  # emtpy set: 0 cost
  if s == 0:
    return 0
  if d == 0:
    # check if s is a power of 2, meaning it contains a single item
    # alternatives:
    #   * Python >= 3.10: s.bit_count() == 1
    #   * Python <  3.10: bin(s).count('1') == 1
    if (s & (s - 1)) == 0:
      return price[int(log2(s))][0]
    # else, if multiple items in the set, we can only buy 1
    best = 2e9
    for x in range(k):
      if (s & (1 << x)) and (price[x][0] < best):
        best = price[x][0]
    return best

  # either don't buy a product at day d
  best = solve_rec(s, d - 1)
  # or buy a product that belongs to s at day d
  for x in range(k):
    if s & (1 << x):
      best = min(best, solve_rec(s ^ (1 << x), d - 1) + price[x][d])

  return best


# total[s][d]: min price of buying subset s at day d
total = [[2e9] * n for _ in range(1 << k)]


# O(2^k * n * k)
# Same with recursive, but significantly faster due to memoization.
def solve_dp():
  # empty set costs nothing at any day
  for d in range(n):
    total[0][d] = 0

  # we can only pick one item at the 0th day
  for s in range(1, 1 << k):
    if (s & (s - 1)) == 0:
      # single-item sets
      total[s][0] = price[int(log2(s))][0]
    else:
      # pick the cheapest item from multi-item sets
      for x in range(k):
        if (s & (1 << x)) and (price[x][0] < total[s][0]):
          total[s][0] = price[x][0]

  for d in range(1, n):
    for s in range(1, 1 << k):
      best = total[s][d - 1]
      for x in range(k):
        if s & (1 << x):
          best = min(best, total[s ^ (1 << x)][d - 1] + price[x][d])
      total[s][d] = best
  return total[-1][-1]


if __name__ == '__main__':
  cost_rec = solve_rec((1 << k) - 1, n - 1)
  cost_dp = solve_dp()
  print("total cost:", cost_dp)
  assert cost_rec == cost_dp